go-controller/pkg/util/node_annotations_unit_test.go

```diff
@@ -0,0 +1,298 @@
+			res, e := tc.inpL3GwCfg.MarshalJSON()
+			t.Log(string(res), e)
+			assert.True(t, reflect.DeepEqual(res, tc.expOutput))
+		})
+	}
+}
+
+func TestL3GatewayConfig_UnmarshalJSON(t *testing.T) {
+	tests := []struct {
+		desc        string
+		expOut      L3GatewayConfig
+		inputParam  []byte
+		errExpected bool
+	}{
+		{
+			desc:        "error: test bad input causing json Unmarshal error",
+			errExpected: true,
+			inputParam:  []byte(`{`),
+		},
+		{
+			desc:       "success: test gateway mode disabled path",
+			inputParam: []byte(`{"mode":""}`),
+			expOut: L3GatewayConfig{
+				Mode:           "",
+				NodePortEnable: false,
+			},
+		},
+		{
+			desc:        "error: test unsupported gateway mode",
+			inputParam:  []byte(`{"mode":"blah"}`),
+			errExpected: true,
+		},
+		{
+			desc:        "error: test bad VLANID input",
+			errExpected: true,
+			inputParam:  []byte(`{"mode":"local","vlan-id":"A"}`),
+		},
+
+		{
+			desc:        "test bad MAC address value",
+			errExpected: true,
+			inputParam:  []byte(`{"mode":"local","mac-address":"BADMAC"}`),
+		},
+		{
+			desc:        "test bad 'IP address' value",
+			errExpected: true,
+			inputParam:  []byte(`{"mode":"local","mac-address":"11:22:33:44:55:66","ip-address":"192.168.1/24"}`),
+		},
+		{
+			desc:        "test bad 'IP addresses' value",
+			errExpected: true,
+			inputParam:  []byte(`{"mode":"local","mac-address":"11:22:33:44:55:66","ip-addresses":["192.168.1/24","fd01::1234/64"]}`),
+		},
+		{
+			desc:        "test bad 'next-hop value",
+			errExpected: true,
+			inputParam:  []byte(`{"mode":"local","mac-address":"11:22:33:44:55:66","ip-address":"192.168.1.5/24","next-hop":"192.168.11"}`),
+		},
+		{
+			desc:        "test bad 'next-hops' value",
+			errExpected: true,
+			inputParam:  []byte(`{"mode":"local","mac-address":"11:22:33:44:55:66","ip-address":"192.168.1.5/24", "next-hops":["192.168.1.","fd01::1"]}`),
+		},
+	}
+	for i, tc := range tests {
+		t.Run(fmt.Sprintf("%d:%s", i, tc.desc), func(t *testing.T) {
+			l3GwCfg := L3GatewayConfig{}
+			e := l3GwCfg.UnmarshalJSON(tc.inputParam)
+			if tc.errExpected {
+				t.Log(e)
+				assert.Error(t, e)
+			} else {
+				t.Log(l3GwCfg)
+				assert.Equal(t, l3GwCfg, tc.expOut)
+			}
+		})
+	}
+}
+
+func TestSetL3GatewayConfig(t *testing.T) {
+	fakeClient := fake.NewSimpleClientset(&v1.NodeList{})
+	k := &kube.Kube{KClient: fakeClient}
+	testAnnotator := kube.NewNodeAnnotator(k, &v1.Node{})
+
+	tests := []struct {
+		desc             string
+		inpNodeAnnotator kube.Annotator
+		inputL3GwCfg     L3GatewayConfig
+		errExpected      bool
+	}{
+		{
+			desc:             "success: empty L3GatewayConfig applied should pass",
+			inpNodeAnnotator: testAnnotator,
+			inputL3GwCfg:     L3GatewayConfig{},
+		},
+		{
+			desc:             "success: apply empty Chassis id ",
+			inpNodeAnnotator: testAnnotator,
+			inputL3GwCfg: L3GatewayConfig{
+				ChassisID: " ",
+			},
+		},
+	}
```

Comment:
This is clearly not a complete set of tests for `SetL3GatewayConfig`...

Reply:
Let me relook into the function code and see what I may have missed. Seems like I only have success cases currently.

Reply:
![Screen Shot 2020-07-02 at 4 41 34 PM](https://user-images.githubusercontent.com/62150532/86411659-f5925600-bc82-11ea-9de9-9eafc1adb330.png)

From the coverage report, I am seeing that I am missing error path testing. I am guessing that's probably what you maybe referring to. I shall create the mocks for the Annotator interface in pkg/kube to test the error path

Reply:
well, I was thinking of `SetL3GatewayConfig` as being the function that did the marshalling, but I guess you effectively test that via the `MarshalJSON` test

Reply:
> well, I was thinking of `SetL3GatewayConfig` as being the function that did the marshalling, but I guess you effectively test that via the `MarshalJSON` test

Yes, MarshalJSON() code path is 100% covered as part of the unit test
![Screen Shot 2020-07-07 at 10 09 17 AM](https://user-images.githubusercontent.com/62150532/86801977-2d5e1c80-c03a-11ea-9efe-7e6b18be0f3c.png)

Reply:
Additionally, I have added tests to cover the error paths for the `SetL3GatewayConfig` function
![Screen Shot 2020-07-07 at 10 14 32 AM](https://user-images.githubusercontent.com/62150532/86802637-d4db4f00-c03a-11ea-918e-7020217ac679.png)